File: Monotonic/max_sliding_window.py
from collections import deque
from typing import List


class MaxSlidingWindow:
    def max_sliding_window(self, nums: List[int], k: int) -> List[int]:
        windowMaxes = []
        monoDecreasingDeque = deque()
        
        for i in range(len(nums)):
            # Remove indices from the front of the deque if they are out of the current window
            if monoDecreasingDeque and monoDecreasingDeque[0] < i - k + 1:
                monoDecreasingDeque.popleft()

            # Pop indices from the deque while the current number is greater than the number at those indices
            while monoDecreasingDeque and nums[i] > nums[monoDecreasingDeque[-1]]:
                monoDecreasingDeque.pop()

            monoDecreasingDeque.append(i)

            # Only start adding to the output array once we have a valid window of size k
            if i >= k - 1:
              windowMaxes.append(nums[monoDecreasingDeque[0]])

        return windowMaxes


# Time Complexity: O(n) where n is the number of elements in the input array nums, as each element is processed at most twice (once when added to the deque and once when removed).
# Space Complexity: O(k) in the worst case, when the deque contains all the indices of the current window. The output array windowMaxes also takes O(n) space.

File: Monotonic/maxSlidingWindow.js
/*You are given an array of integers nums, 
there is a sliding window of size k which is moving from the very left of the array to the very right. 
You can only see the k numbers in the window. 
Each time the sliding window moves right by one position.

Return the max sliding window.*/

const maxSlidingWindow = (nums, k) => {
    const maxDeque = [];
    const ans = new Array(nums.length - k + 1);

    for (let i = 0; i < nums.length; i++) {
        //maintain monotonic decreasing deque and
        //remove all elements in deque smaller than current one (no chance of being maximum)
        while (maxDeque.length && nums[i] > nums[maxDeque.at(-1)]) {
            maxDeque.pop();
        }

        maxDeque.push(i);

        //queue[0] is index of maximum element
        //check if outside window, and remove if so
        if (maxDeque[0] < i - k + 1) {
            maxDeque.shift();
        }

        //add to answer once window is size k
        if (i >= k - 1) {
            ans[i - k + 1] = nums[maxDeque[0]];
        }
    }

    return ans;
}

const nums = [1], k = 1;
console.log(maxSlidingWindow(nums, k));